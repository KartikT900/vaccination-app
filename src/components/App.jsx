import React, { useEffect, useState } from 'react';
import { hot } from 'react-hot-loader/root';

import tt from '@tomtom-international/web-sdk-services';

import AppContainer from 'containers/AppContainer';

export const TT_SERVICE_API_KEY = process.env.TT_KEY || 'key';

function App() {
  const [isGeoEnabled, setIsGeoEnabled] = useState(false);
  const [isGeoPermissionDenied, setGeoPermissionDenied] =
    useState(false);
  const [coordinates, setCoordinates] = useState(null);
  const [localeData, setAreaLocaleData] = useState(null);
  const [fetchingGeoData, setFetchingGeoData] = useState(true);

  const geoSuccessHandler = (position) => {
    const latitude = position.coords.latitude;
    const longitude = position.coords.longitude;

    setCoordinates({
      latitude,
      longitude
    });
    setIsGeoEnabled(true);

    return;
  };

  const geoErrorHandler = () => setGeoPermissionDenied(true);

  useEffect(() => {
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition(
        geoSuccessHandler,
        geoErrorHandler
      );
    } else {
      setIsGeoEnabled(false);
    }
  }, []);

  useEffect(() => {
    async function loadLocaleData() {
      try {
        const arealocaleData = await tt.services.reverseGeocode({
          key: TT_SERVICE_API_KEY,
          position: coordinates
        });

        setAreaLocaleData(arealocaleData);
        setFetchingGeoData(false);
      } catch (error) {
        console.error(error);
        setFetchingGeoData(false);
      }
    }

    if (coordinates && isGeoEnabled) {
      loadLocaleData();
    }
  }, [coordinates, isGeoEnabled]);

  return (
    <AppContainer
      isGeoEnabled={isGeoEnabled}
      isGeoPermissionDenied={isGeoPermissionDenied}
      localeData={localeData}
      geoDataLoading={fetchingGeoData}
    />
  );
}

export default hot(App);
