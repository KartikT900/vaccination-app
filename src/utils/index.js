import axios from 'axios';
import isEmpty from 'lodash/isEmpty';

/**
 *
 * @param {string} url A string value of the resource that will be fetched
 * @param {object} options An object listing out various operational parameters required for fetch request to begin
 * @returns Fetched response
 */
export async function netCall(baseUrl, url, options) {
  if (typeof url !== 'string' || !url) {
    console.error(
      `Expected url to be a string. Instead received ${typeof url}`
    );

    return null;
  }

  axios.defaults.baseURL = baseUrl;

  const { method, headers, data } = options;

  const axiosConfig = {
    method,
    url,
    data: data || '',
    headers: headers || ''
  };

  const response = await axios({ ...axiosConfig });

  return response.data;
}

/**
 *
 * @param  {...any} args Input of classNames to be combined
 *
 * A simple program that returns a string of joined classNames.
 * @example
 * // returns 'test-class test-class2 test-class3'
 * classnames('test-class', 'test-class2', 'test-class3');
 */

export const classnames = (...args) => {
  const classes = args.map((arg) => {
    if (!arg) {
      return;
    } else if (typeof arg === 'number') {
      throw new Error(`Expected some input`);
    } else {
      return arg;
    }
  });

  const classArray = classes.reduce((acc, curr) => {
    if (curr?.constructor === Object) {
      const classString = Object.values(curr)[0]
        ? Object.keys(curr)[0]
        : null;

      acc.push(classString);
    } else {
      acc.push(curr);
    }

    return acc;
  }, []);

  const joinedClasses = classArray
    .filter((value) => value)
    .join(' ')
    .trim();

  return joinedClasses;
};

export const filterDataByKey = (data = [], key) => {
  if (isEmpty(data)) {
    return null;
  } else if (key && key.length === 0) {
    return data;
  }

  const checkKeyExists = (item) =>
    key.includes(item?.vaccineName?.toLowerCase());

  return data?.filter(checkKeyExists);
};
