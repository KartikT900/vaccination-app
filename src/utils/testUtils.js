import React from 'react';
import { render } from '@testing-library/react';

export const withContextProvider = (
  LocalContext,
  Ui,
  { providerProps, ...renderOptions }
) => {
  return render(
    <LocalContext.Provider {...providerProps}>
      {<Ui />}
    </LocalContext.Provider>,
    renderOptions
  );
};
