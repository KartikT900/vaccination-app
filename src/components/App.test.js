import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';

import App from 'components/App';

import * as utils from 'utils';

describe('<App />', () => {
  it('renders correctly', async () => {
    const netCallSpy = jest.spyOn(utils, 'netCall');

    netCallSpy.mockReturnValue({});

    render(<App />);

    expect(screen.getByText('fetching data now')).toBeInTheDocument();

    await waitFor(() => screen.getByText('data: {}'));

    expect(screen.getByText('data: {}')).toBeInTheDocument();
  });
});
