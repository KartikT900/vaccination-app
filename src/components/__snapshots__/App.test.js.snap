// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="vcc-app-container"
  >
    <div
      class="ui active transition visible dimmer"
      style="display: flex;"
    >
      <div
        class="content"
      >
        <div
          class="ui loader"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
